byzcoin/api.go

```diff
@@ -266,6 +266,34 @@ func (c *Client) StreamTransactions(genesisID skipchain.SkipBlockID, handler fun
 	}
 }
 
+// DownloadState is used by a new node to ask to download the global state.
+// The first call to DownloadState needs to have start = 0, so that the
+// service creates a snapshot of the current state which it will serve over
+// multiple requests.
+//
+// Every subsequent request should have start incremented by 'len'.
+// If start > than the number of StateChanges available, an empty slice of
+// StateChanges is returned.
+//
+// If less than 'len' StateChanges are available, only the remaining
+// StateChanges are returned.
+//
+// The first StateChange with start == 0 holds the metadata of the
+// trie which can be `protobuf.Decode`d into a struct{map[string][]byte}.
+func (c *Client) DownloadState(byzcoinID skipchain.SkipBlockID, reset bool, length int) (reply *DownloadStateResponse, err error) {
+	if length <= 0 {
+		return nil, errors.New("invalid parameter")
+	}
+
+	reply = &DownloadStateResponse{}
+	err = c.SendProtobuf(c.Roster.List[len(c.Roster.List)/2], &DownloadState{
```

Comment:
Yes, true. But we should
- exclude the leader, to avoid overloading him
- exclude the last entries in the roster, because they are the most probable to be out of sync (due to a viewchange)

So I went for the middle of the list... Any proposition is welcome.

Reply:
Then a random pick over [1:n/2] so that it takes the first half without the leader ?

Reply:
Still starting with the first non-subleader, but then going through all nodes. I think it's better to avoid the last elements as much as possible, in case they are defective.